### ISSUE TITLE
undefined __snapshot property creates Possible Unhandled Promise Rejection error on set

### ISSUE THREAD

**Issue by estebanrao** [nandorojo/swr-firestore#130]
Taking this as an example:

`const { data, set } = useDocument(`myCollection/${userId}`, { listen: true })`

`data` comes with `__snapshot: undefined` property, if I use this data as state for, let's say a form:

```
useEffect(() => {
    if (data) {
      setValues({ ...data, otherProp: otherValue });
```

If I wan't to do:

```
    set(
      values
      { merge: true }
    )
```

I'm getting this error:
```
Possible Unhandled Promise Rejection: FirebaseError: Function DocumentReference.set() called with invalid data. Unsupported field value: undefined (found in field __snapshot in document [...])
```

Shouldn't we just avoid the empty `__snapshot` property at all? or is this a feature and I should just sanitize my data object?

**Comment by estebanrao** [comment]
I've checked and I can use `data.__snapshot.data()` for the exact data object I was accessing through `doc.data()` before once I specify `ignoreFirestoreDocumentSnapshotField: false,`. Is this the "safe" way of accessing my data for modification?

**Comment by linuz90** [comment]
I'm having the same issue. 

After getting the data with `useDocument`, I can't set it with `set(data)` because I get that error, even if I set `ignoreFirestoreDocumentSnapshotField: true` in the `useDocument`.

**Comment by estebanrao** [comment]
@linuz90 check my second comment in this thread. I think that the confusion in mainly because of naming, you'd expect `data` to be equal to `snapshot.data()` and not the document itself. Maybe it would have been better to have 
const `{ doc, ..., data (as in snapshot.data()) } = useDocument(...)` ?

**Comment by nandorojo** [comment]
Maybe this field should be set with `Object.defineProperty` with a getter such that it isn't enumerable. Would that solve it?

**Comment by estebanrao** [comment]
Hey @nandorojo ! Well... `__snapshot` is not the only property that will misbehave on `set` cause data also comes with additional properties that are internal to firestore, like `exists` or `id`. As I mentioned in my second comment, I was able to get data correctly by doing `data.__snapshot.data()` but it feels hacky, I guess someone would expect `data` to equal `.data()` in firestore and then maybe have a `snapshot` property in `useDocument` and `useCollection`. Not a big deal actually by maybe something that should be documented? What do you think?

**Comment by nandorojo** [comment]
I believe this is all documented

**Comment by estebanrao** [comment]
Yeah... maybe it was a rookie mistake using the library. I still think this particular case is not covered in the documentation: get data, work with that data, update firestore. Which is a very common use case. Nothing that you can solve checking the rest of the docs though.

BTW, great work with this library!

**Comment by nandorojo** [comment]
I’d say for that use case, the `__snapshot.data()` covers it. Form editing is a more unique use case to need the exact data for, which is why that field exists. We could add that more explicitly to the readme

**Comment by nandorojo** [comment]
> I'm having the same issue.
> 
> After getting the data with `useDocument`, I can't set it with `set(data)` because I get that error, even if I set `ignoreFirestoreDocumentSnapshotField: true` in the `useDocument`.

Try `set(data.__snapshot.data())`, and set the ignore field to false.

**Comment by estebanrao** [comment]
> I’d say for that use case, the `__snapshot.data()` covers it. Form editing is a more unique use case to need the exact data for, which is why that field exists. We could add that more explicitly to the readme

Sounds good! I can take care of that if you'd like

**Comment by nandorojo** [comment]
Sure! Thanks

**Comment by nandorojo** [comment]
One example would be, if you want to set `initialValues` on formik, use the snapshot data
